app/src/test/java/supermarket/ProductTest.kt

```diff
@@ -1,84 +1,303 @@
-        cart.addItemQuantity(toothbrush, 1.0)
+    @After
+    fun tearDown() {
+        // No teardown necessary, but print receipt while we're at it.
+        receipt?.let {
+            println(ReceiptPrinter().printReceipt(receipt!!))
+        }
+    }
 
+    @Test
+    fun `Require items in cart to apply discounts`() {
         teller.addSpecialOffer(PercentageOffer(toothbrush, 10.0))
 
-        val receipt = teller.checksOutArticlesFrom(cart)
-        val discounts = receipt.getDiscounts()
-        assertThat(discounts.size, Is(1))
+        receipt = teller.checksOutArticlesFrom(cart)
+        val discounts = receipt!!.getDiscounts()
+
+        assertThat(discounts.size, Is(0))
     }
 
     @Test
-    fun applesDiscount() {
-        cart.addItemQuantity(apples, 2.0)
-        teller.addSpecialOffer(PercentageOffer(apples, 50.0))
-        val receipt = teller.checksOutArticlesFrom(cart)
-        val discounts = receipt.getDiscounts()
+    fun `Require special offers to apply discounts`() {
+        cart.addItemQuantity(toothbrush, 25.0)
+
+        receipt = teller.checksOutArticlesFrom(cart)
+        val discounts = receipt!!.getDiscounts()
+
+        assertThat(discounts.size, Is(0))
+    }
+
+    @Test
+    fun `Discounts are only applied to relevant products`() {
+        cart.addItemQuantity(oranges, 2.0)
+        cart.addItemQuantity(apples, 1.0)
+        teller.addSpecialOffer(PercentageOffer(oranges, 25.0))
+        teller.addSpecialOffer(ThreeForTwoOffer(toothbrush))
+
+        receipt = teller.checksOutArticlesFrom(cart)
+        val discounts = receipt!!.getDiscounts()
+
         assertThat(discounts.size, Is(1))
-        assertThat(discounts[0].discountAmount, Is(1.99))
+        assertThat(discounts[0].description, containsString("oranges"))
+    }
+
+    @Test
+    fun `Apply multiple discounts for different products`() {
+        cart.addItemQuantity(oranges, 5.0)
+        cart.addItemQuantity(apples, 3.0)
+        teller.addSpecialOffer(PercentageOffer(apples, 10.0))
+        teller.addSpecialOffer(QuantityForAmountOffer(oranges, 4, 5.0))
+
+        receipt = teller.checksOutArticlesFrom(cart)
+        val discounts = receipt!!.getDiscounts()
+
+        assertThat(discounts.size, Is(2))
+        assertThat(discounts[0].description, containsString("apples"))
+        assertThat(discounts[0].discountAmount, Is(0.60))
+        assertThat(discounts[1].description, containsString("oranges"))
+        assertThat(discounts[1].discountAmount, Is(3.00))
+    }
+
+    @Test
+    fun `Apply multiple discounts for same product`() {
+        cart.addItemQuantity(oranges, 5.0)
+        teller.addSpecialOffer(PercentageOffer(oranges, 10.0))
+        teller.addSpecialOffer(QuantityForAmountOffer(oranges, 4, 5.0))
+
+        receipt = teller.checksOutArticlesFrom(cart)
+        val discounts = receipt!!.getDiscounts()
+
+        assertThat(discounts.size, Is(2))
+        assertThat(discounts[0].description, containsString("oranges"))
```

Comment:
We see this pattern all over the place in these tests, seems like some potential for DRY - I was thinking maybe extracting a method like `checkDescriptionAndValue()` but I think just avoiding the repeated index access might be nice enough:
```
        discounts[0].run { 
            assertThat(description, containsString("oranges"))
            assertThat(discountAmount, Is(1.00))
        }
```
Also, I think it's worth either using `oranges.name` (I think that's the best option) or at the least using a constant for the string.